deactivated_user: Change the empty PM / group PM view text and compose_box.

palashb01 opened the pull request:
**If the user is deactivated**:

- change empty PM/group PM view text.
- keep the compose_box closed and blocked for the PM with deactivated user or a group containing a deactivated user.
- block the compose_box in `all_message_narrow` and `recent conversations narrow` if the recipient of the selected message is deactivated.
- Change the text to:

       - pm-with:
         Line 1: You have no direct messages with {person}
         Line 2: [user's name] is no longer active in this organization.

      - group pm-with:
         Line 1: You have no direct messages with these users.
         Line 2: [user's name] is no longer active in this organization.

      - dm-including:
        Line 1: You have no direct messages including {person}.
        Line 2: [user's name] is no longer active in this organization.

- Keep the compose box as-is, rather than opening it up for the "pm-with:" search, as we currently do.
- Change the text of user status tooltip to `User is deactivated`

----------------------------------------

**Fixes** #23408. **Fixes** #22278.
**CZO:** [Thread](https://chat.zulip.org/#narrow/stream/101-design/topic/pm.20with.20deactivated.20user.20behaviour.20.2323899/near/1482943)

----------------------------------------
- **King Hamlit  is deactivated**

### DEMO: 

<details>
<summary>Before:</summary>
<br>

https://user-images.githubusercontent.com/66828942/208317489-803e886e-01e3-4157-8fd1-e0f2c3180f15.mp4
</details>

 

### After:

---------------------------------------------------

<details>
<summary>All message narrow:</summary>
<br>

![all_messages_narrow](https://user-images.githubusercontent.com/66828942/234410788-e95572c7-27e8-4cf5-bce7-b9d33a54b95f.gif)



</details>


<details>
<summary>pm-with-deactivated-user:</summary>
<br>


![pm_with_deactivated_user](https://user-images.githubusercontent.com/66828942/234410815-48edc420-3977-4d43-bf30-0b807ce394c6.gif)


</details>


<details>
<summary>group-pm-having-deactivated-user:</summary>
<br>


![group-pm](https://user-images.githubusercontent.com/66828942/234410878-4ff004c9-b987-49d1-bc21-71adf5213c12.gif)


</details>

<details>
<summary>narrow-banner-change:</summary>
<br>

![Screenshot from 2023-04-26 02-41-35](https://user-images.githubusercontent.com/66828942/234410950-dd2fd3b6-5893-41de-bfab-93bdc43c8423.png)



<br>

![Screenshot from 2023-04-26 02-41-43](https://user-images.githubusercontent.com/66828942/234410979-649c7357-7254-4396-a4e4-aa3a71834af5.png)

<br>

![Screenshot from 2023-04-26 03-06-15](https://user-images.githubusercontent.com/66828942/234411018-9a0a202a-d4de-41ab-9a12-2bcb8595c2d1.png)


</details>

<details>
<summary>User status tooltip:</summary>
<br>

![Screenshot from 2023-03-18 06-51-20](https://user-images.githubusercontent.com/66828942/226075726-6e8a59a4-9b3f-4e79-a141-40345e6216b7.png)


</details>

------------------------------

<details>
<summary>Self-review checklist</summary>


- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

palashb01 commented:
Hey @alya @amanagr , Could you please review the PR, and suggest if any changes are required, Thanks.

N-Shar-ma commented on static/js/compose_actions.js:
```diff
@@ -292,10 +292,22 @@ export function start(msg_type, opts) {
 
     opts = fill_in_opts_from_current_narrowed_view(msg_type, opts);
 
+    if (
```
A comment here explaining that we close the compose box when narrowing to a deactivated user, else we expand it, would be nice

N-Shar-ma commented on static/js/compose_actions.js:
```diff
@@ -331,7 +343,6 @@ export function start(msg_type, opts) {
     // explicit "clear" function for compose_state.
     compose_state.set_stream_name(opts.stream);
     compose_state.topic(opts.topic);
-
```
This line break removal is not needed so please restore it

N-Shar-ma reviewed (changes requested):
Thanks for working on this! The code and screencaps mostly look good, have left a couple minor comments. Please fix the failing test though before any more in depth reviews

palashb01 commented:
Hey @N-Shar-ma , I have fixed the failing test and made the appropriate changes in the node_test, could you please review it again and suggest if any changes are required, Thanks.

N-Shar-ma commented on frontend_tests/node_tests/narrow.js:
```diff
@@ -436,6 +436,33 @@ run_test("show_empty_narrow_message", ({mock_template}) => {
+            "translated: You have no private messages with Alice Smith yet.",
+            'translated HTML: Why not <a href="#" class="empty_feed_compose_private">start the conversation</a>?',
+        ),
+    );
```
This block of code is not needed I think, since this was checked immediately after enabling sending private messages

palashb01 replied:
Some of the node_tests are failing without this piece of code, I have attached the log also this was discussed on [CZO](https://chat.zulip.org/#narrow/stream/49-development-help/topic/.E2.9C.94.20node.20test.20for.20.2323899)

![Screenshot from 2022-12-24 23-46-52](https://user-images.githubusercontent.com/66828942/209447750-33bcf2ad-2bfa-4cc7-b685-03878148694d.png)

Please do suggest, if some more changes are required, Thanks.

N-Shar-ma replied:
Did you remove the code both above and below this comment? Sorry if I wasn't clear before, but I meant to only refer to the new code above this comment. In other words, let the new code from `people.deactivate(alice)` stay as is and remove the block right above. The tests should still work then

N-Shar-ma replied:
Looks like you're removing the code blocks both above and below this comment. Remove just the one above. I've tested locally too, that works (node tests don't fail)

palashb01 replied:
Hey @N-Shar-ma ,Made the changes suggested by you, kindly review it  and suggest if any more changes  are required, Thanks.

N-Shar-ma commented on frontend_tests/node_tests/narrow.js:
```diff
@@ -496,6 +523,32 @@ run_test("show_empty_narrow_message", ({mock_template}) => {
+            "translated: You have no group private messages with Raymond yet.",
+            'translated HTML: Why not <a href="#" class="empty_feed_compose_private">start the conversation</a>?',
+        ),
+    );
```
This block of code too seems extraneous for the same reason

palashb01 replied:
The same reason goes for this code snippet too.

N-Shar-ma replied:
Here too, just remove the new code above this comment, not below

N-Shar-ma replied:
Just remove the code block above, not below this comment

N-Shar-ma commented on static/js/compose_actions.js:
```diff
@@ -292,10 +292,29 @@ export function start(msg_type, opts) {
+                    people.emails_strings_to_user_ids_string(opts.private_message_recipient),
+                    10,
+                ),
+            )
```
Un-nesting this `if`, using `&&` should make this code more concise and readable

palashb01 replied:
made the change, Thanks

N-Shar-ma reviewed (approved):
LGTM now!

palashb01 commented:
Hey @alya , I have raised a PR and reviewed this with @N-Shar-ma , could you please have a look at the final draft and suggest if any more changes are required, Thanks.

N-Shar-ma commented:
I played around with the PR locally and have some thoughts... 

All 3 compose buttons in the bottom bar being unresponsive when viewing pms with a deactivated user, seems more like a bug than intentional. I feel the "New stream message" and "New private message" buttons should definitely behave as usual and the "Message [deactivated user-name]" button should be visually shown to be disabled too. Maybe you could create a CZO thread in #design to finalize the exact behaviour before implementing this, @palashb01 
![image](https://user-images.githubusercontent.com/68962290/209488799-912d0c5f-c379-4a98-8edb-50a3f18a994d.png)

palashb01 commented:
> I played around with the PR locally and have some thoughts...
> 
> All 3 compose buttons in the bottom bar being unresponsive when viewing pms with a deactivated user, seems more like a bug than intentional. I feel the "New stream message" and "New private message" buttons should definitely behave as usual and the "Message [deactivated user-name]" button should be visually shown to be disabled too. Maybe you could create a CZO thread in #design to finalize the exact behaviour before implementing this, @palashb01 ![image](https://user-images.githubusercontent.com/68962290/209488799-912d0c5f-c379-4a98-8edb-50a3f18a994d.png)

Okay, will do that, thanks a lot for the review.

N-Shar-ma commented:
@palashb01 Could you link to the CZO thread here so that someone looking through this PR has all the relevant context?

palashb01 commented:
> @palashb01 Could you link to the CZO thread here so that someone looking through this PR has all the relevant context?

Sure, here is the link to that [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/pm.20with.20deactivated.20user.20behaviour.20.2323899/near/1482943)

alya commented:
> Okay, will do that, thanks a lot for the review.

Please post a comment and include and updated screenshot when you've had a chance to address @N-Shar-ma 's feedback. Thanks!

palashb01 commented:
> > Okay, will do that, thanks a lot for the review.
> 
> Please post a comment and include and updated screenshot when you've had a chance to address @N-Shar-ma 's feedback. Thanks!

Sure, working on the Issue will update the PR soon, Thanks.

palashb01 commented:
Hey @alya @N-Shar-ma , I have updated the PR with the latest changes could you please review it once and suggest if any changes are required, Thanks.

```
- Disabled the compose box, when pm-with deactivated user.
- Showing a description on hover over compose box.
- Changed the banner when pm-with deactivated user.
```


**Demo Video:**

https://user-images.githubusercontent.com/66828942/211879786-db89c4a0-7206-41d5-840b-7dab295ad56a.mp4

**Compose box:**
![zulip](https://user-images.githubusercontent.com/66828942/211879931-02203e7a-27d2-4921-8577-ce2a5948f500.gif)

alya commented:
Thanks! It looks good in manual testing, but I noticed that the button does not live-update properly when a user is deactivated or reactivated. It's not a dealbreaker for this feature, but would be good to do.

@amanagr perhaps it would be good for you to take a look next?

amanagr commented on static/js/compose_closed_ui.js:
```diff
@@ -64,17 +64,63 @@ function update_conversation_button(btn_text, title) {
+    if (narrow_state.pm_ids_string()) {
+        const user_ids = people.user_ids_string_to_ids_array(narrow_state.pm_ids_string());
+        const user_ids_dicts = user_ids.map((user_id) => ({id: user_id}));
+        if (people.verify_non_active_human_id(user_ids_dicts[0].id)) {
+            // Disable compose box with a description when the recipient is a deactivated user.
+            $(".compose_reply_button").attr("disabled", true);
+            const disable_reply = true;
+            const title_reply = $t({
+                defaultMessage: "Cannot send message to a deactivated user.",
+            });
+            update_buttons(text_stream, disable_reply, title_reply);
+        } else {
+            update_buttons(text_stream);
+        }
```
Is this a real case? Deactivated users shouldn't be able to login, that would defeat the purpose of them being deactivated.

palashb01 replied:
Yes, deactivated users are not allowed to log in, but in this case, an active user is trying to message a deactivated user and we are blocking the ability to send a message to a deactivated user.

- A active user is logged in.
- pm-with deactivated user.

amanagr replied:
oh yeah!

amanagr commented on static/js/compose_closed_ui.js:
```diff
@@ -64,17 +64,63 @@ function update_conversation_button(btn_text, title) {
     const title_conversation = text_conversation + " (x)";
+    const text_deactivated_conversation = $t({defaultMessage: "Reply selected message"});
+    let title_deactivated_conversation = text_deactivated_conversation + " (r)";
+    if (narrow_state.pm_ids_string()) {
+        const user_ids = people.user_ids_string_to_ids_array(narrow_state.pm_ids_string());
+        const user_ids_dicts = user_ids.map((user_id) => ({id: user_id}));
+        if (people.verify_non_active_human_id(user_ids_dicts[0].id)) {
```
This could be a group PM too, right? In that case, we should check if any of the users are deactivated and give an appropriate message stating exactly who is deactivated. Also, we don't need to call `narrow_state.pm_ids_string()` twice here.

amanagr commented on static/js/compose_closed_ui.js:
```diff
@@ -122,10 +168,16 @@ export function initialize() {
     // Click handlers for buttons in the compose box.
     $("body").on("click", ".compose_stream_button", () => {
         compose_actions.start("stream", {trigger: "new topic button"});
+        // When pm-with deactivated user compose box remains closed,
+        // however if a user clicks on new stream message or private message buttons,
+        // both the buttons should work as intended, thus initialize expand_compose_box function
+        // to open compose_box.
+        compose_actions.expand_compose_box();
     });
 
     $("body").on("click", ".compose_private_button", () => {
         compose_actions.start("private", {trigger: "new private message"});
+        compose_actions.expand_compose_box();
```
You should check for the `trigger` inside `compose_actions.start` to verify if you should expand the compose box, instead of calling `expand_compose_box` in a different line.

palashb01 replied:
Sure will look into all the suggested changes

amanagr commented on static/js/compose_actions.js:
```diff
@@ -292,10 +306,32 @@ export function start(msg_type, opts) {
+    // the recipient is a deactivated user else,
+    // expand compose box.
+    if (
+        opts.private_message_recipient !== "" &&
+        check_pm_deactivated(
+            people.emails_strings_to_user_ids_string(opts.private_message_recipient),
+        ).is_user_id_deactivated
+    ) {
+        // Expand compose box with a deactivated user also,
+        // if it is the case with stream or new private message button.
+        if (
+            (msg_type === "stream" && opts.trigger === "new topic button") ||
+            (msg_type === "private" && opts.trigger === "new private message")
```
Let's combine these two `if` conditions. Maybe add a `check_new_top_or_pm_trigger()` to simplify reading it.

Also, we should update the compose box deactivated state once a user is deactivated or activated back.

amanagr replied:
Updated the PR with what I was thinking here.

palashb01 replied:
Thanks.

amanagr commented:
@palashb01 thanks! Left a comment.

amanagr commented:
Also, this doesn't seem to be working when the user is not all message narrow.

amanagr commented:
I am not sure about the documentation error, maybe just rebasing it on `main` will fix it.

palashb01 commented:
> 



> Also, this doesn't seem to be working when the user is in all message narrow.

Ohh, looking into it thanks for the review.

palashb01 commented:
Hey, @alya @amanagr, could you kindly examine the PR and recommend any changes? I have updated the **PR description** with the most recent screenshots and changes, and I have provided a brief commit message summarising each change/function.

amanagr committed:
popovers: Fix reactivate user button.

Since reactivate user button isn't present inside `.info_popover_actions`
in the HTML DOM the click handler was not working.

amanagr commented:
@palashb01 thanks for the update! I pushed the PR with minor changes and added a commit to fix the reactivate user button. You can fetch these changes using `./tools/fetch-rebase-pull-request 23408`. 

The compose box is not being updated when I deactivate a user or reactive it when already in a PM with the user. This is the last bit of feature we feel important to fix.

palashb01 commented:
> @palashb01 thanks for the update! I pushed the PR with minor changes and added a commit to fix the reactivate user button. You can fetch these changes using `./tools/fetch-rebase-pull-request 23408`.
> 
> The compose box is not being updated when I deactivate a user or reactive it when already in a PM with the user. This is the last bit of feature we feel important to fix.

Sure looking into it, thanks for the review.

amanagr commented:
Pushed to fix the 1st commit message a bit.

amanagr commented:
We should also update the right sidebar once the user is activated again.

amanagr commented:
> Should I create a different PR for this Issue or work on this one only, I believe it would be much better to work on this issue in a separate PR.

It is most likely  3-4 lines of change, you can open a PR if it feels like a huge change to you.

palashb01 commented:
Hey @amanagr, I have made the changes to update the compose_box instantly when a user is deactivated without reload. However, I am getting node_tests fail in `node_tests/example4.js`. Could you please have a look at the final PR and suggest a way for the node_tests? I tried to override it, however, it didn't work out. CZO thread : [Link](https://chat.zulip.org/#narrow/stream/49-development-help/topic/disable.20compose.20box-user.20is.20deactivated)


![zulip](https://user-images.githubusercontent.com/66828942/215274699-0aedd7f6-19a8-4f1e-b973-bcae98ad0345.gif)

@alya I have updated the user_status popover, could you please have a look at it, Thanks.

![Screenshot from 2023-01-28 20-46-57](https://user-images.githubusercontent.com/66828942/215274734-323a437b-961d-4d1f-b809-512b60906519.png)

alya commented:
The tooltip screenshot looks good, thanks! I'll retest once the code is ready.

palashb01 commented:
Hey @alya @amanagr , the PR is ready for manual testing and code review, could you please review the PR, thanks.

amanagr commented:
@palashb01 LGTM, you should squash the last two commits.

palashb01 commented:
> @palashb01 LGTM, you should squash the last two commits.
 
Sure

palashb01 commented:
Hey @amanagr , I have squashed all the commits into one.

alya commented:
@palashb01 The commits still don't look quite right:
<img width="578" alt="Screen Shot 2023-01-31 at 11 11 18 PM" src="https://user-images.githubusercontent.com/2090066/215975674-c2660029-5513-4157-b86a-8079eed1585d.png">

palashb01 commented:
> @palashb01 The commits still don't look quite right: <img alt="Screen Shot 2023-01-31 at 11 11 18 PM" width="578" src="https://user-images.githubusercontent.com/2090066/215975674-c2660029-5513-4157-b86a-8079eed1585d.png">

@alya, I have squashed both commits. The second one appeared due to a merge conflict that I resolved using the GitHub GUI. However, I have now made the changes to show only one commit without conflicts. Could you please review the PR, Thank you.

amanagr commented:
@palashb01 to be clear there should be two commits here, one which fixes the reactive user button in user popover (which I pushed to the PR earlier) and the second one which includes all your changes.

palashb01 commented:
@amanagr I squashed all the commits into one using `git rebase -i Head~` , and now I am not quite sure how to revert back to the old commits. Could you please suggest a way for me to do that.

palashb01 commented:
> @palashb01 to be clear there should be two commits here, one which fixes the reactive user button in user popover (which I pushed to the PR earlier) and the second one which includes all your changes.

Hey @amanagr  , I have reverted the commits so that there are now two separate commits, one from you and one from me. Could you please review the PR, Thanks.

amanagr commented:
@palashb01 can you rebase this please?

palashb01 commented:
> @palashb01 can you rebase this please?

Hey @amanagr , I have rebased the PR, could you please review the PR, Thanks.

amanagr commented:
LGTM!

alya commented:
Thanks!

1. Can we switch to Tippy tooltips for the deactivated "Message..." button? 
2. The `x` keyboard shortcut should still work when you are viewing messages with a deactivated user.

alya commented:
I also noticed that the text (both lines) doesn't actually match the spec in the issue. Please be sure to check carefully next time, and post an explicit comment if you'd like to discuss a potential change.

Line 1: You have no {private messages/group private messages} with this person.
Line 2: [user's name] is no longer active in this organization.

<img width="485" alt="Screen Shot 2023-02-07 at 12 49 18 AM" src="https://user-images.githubusercontent.com/2090066/217196555-fa1da56c-f427-4052-88f2-611246d3a8c1.png">

alya commented:
Also, the empty message feed message for group PMs that include a deactivated user has not been implemented:
<img width="905" alt="Screen Shot 2023-02-07 at 12 52 21 AM" src="https://user-images.githubusercontent.com/2090066/217197429-60cff9bd-e0d9-47a2-bb3e-966a3cca01fb.png">

palashb01 commented:
> Also, the empty message feed message for group PMs that include a deactivated user has not been implemented: <img alt="Screen Shot 2023-02-07 at 12 52 21 AM" width="905" src="https://user-images.githubusercontent.com/2090066/217197429-60cff9bd-e0d9-47a2-bb3e-966a3cca01fb.png">

Working on all the suggested changes.

palashb01 commented:
Hey @alya , I've made all the changes you suggested, including the x-keyboard shortcut. However, I think the compose-box tooltips can be migrated to Tippy.js in a separate follow-up issue. Could you please review the PR, Thanks.

alya commented:
Thanks! Sounds good about the tooltips. Could you please update the PR description with the latest set of screenshots, including still (non-animated) screenshots for the changes that address the feedback above?

palashb01 commented:
> Thanks! Sounds good about the tooltips. Could you please update the PR description with the latest set of screenshots, including still (non-animated) screenshots for the changes that address the feedback above?

While playing around with the PR, I realized that there is one edge case not included yet. In the case of a direct message with a deactivated user, the compose-box was not going into the disabled state on recent conversations narrow. I am currently working on this edge case as well and will update the PR once all cases are covered. Thanks.

palashb01 committed:
people: Add functions to obtain and validate deactivated users.

This commit introduces two functions within people.ts. The first
function is used to verify if a given user_id corresponds to a
deactivated user, and it returns a boolean value accordingly. The
second function returns the full name of the first deactivated user
found in an array of user_ids.

palashb01 committed:
tooltip: Update the user status tooltip for deactivated users.

This commit updates the user status tooltip for deactivated users to
a title saying 'User is deactivated'

palashb01 committed:
compose_box: Update the compose box for deactivated users.

This commit updates the compose box for deactivated users.

There are three different locations where the compose box should
be updated for deactivated users:

- PM & group PM
- Recent conversations narrow
- All messages narrow

In all the above cases, the recipient label inside the compose box
changes, which is the key starting point to implement this feature.

To ensure the compose box does not open when clicked on in the case
of a deactivated user, a function 'check_pm_deactivated' was added
inside the 'compose_actions.js' file. Since there is no 'opts' to
determine the recipients for a closed_compose_box, this function
takes two arguments and is required for updating the
closed_compose_box as well. This function checks if the given
user_ids array contains any deactivated users. If so, it returns two
values: the first is a boolean value indicating whether there is a
deactivated user or not, and the second indicates the number of
deactivated users, which is helpful for updating the tooltip
accordingly. Even if the user is deactivated and the compose_box is
disabled, the other two buttons 'New Topic' and 'New Private Message'
should still work, as checked by the 'check_new_topic_or_pm_trigger'
function.

To update the compose box for deactivated users, a function
'update_reply_button_deactivated_users' was added inside
'compose_closed_ui.js'. This function is used to verify if there are
any deactivated users among the recipients and update the compose box
accordingly. To get the recipients, there are three different
approaches for the three cases defined above:

- pm-with: narrow_state contains the pm_ids, which are used to get
            the recipients

- group-pm-with & all messages narrow: the selected message row can be
            used, and to distinguish between stream and pm messages,
            the 'is_Array' function is used, since for stream
            messages, the 'selected_message().display_recipient' is a
            string, but for pm messages, it is an array

- recent conversations narrow: the focused row message is used to get
            the recipients, and to distinguish between stream and pm
            messages, one argument message is passed, which has the
            'display_reply_to' attribute for all pm messages and the
            attribute is 'stream' for stream messages.

To live-update the compose box whenever a user is deactivated or
activated, the server_events_dispatch is used, which simultaneously
updates when the user is deactivated or activated.

Node tests are added according to the changes made.

Fixes #22278

palashb01 commented:
Hey @alya , I have fixed all the edge cases and tested all the changes thoroughly. I've also updated the PR description. Could you please review the PR? Thanks :)

alya commented:
Great, works for me in manual testing (for fixing both issues).

@timabbott , this PR has been reviewed by @N-Shar-ma , @amanagr and me -- please take a look.

timabbott commented on web/src/people.js:
```diff
@@ -926,6 +926,23 @@ export function get_bot_ids() {
+    return false;
+}
+
+export function get_deactivated_users_name(user_id) {
```
The parameter should be called `user_ids`, should it appears to expect a list? 

Also the function name seems wrong, it should be `get_names_for_deactivated_users` or something like that, to hint the fact it expects a list.

palashb01 replied:
@timabbott Thanks for the suggestion, I have changed the name of the parameter and function, could you please review the PR, thanks :)

palashb01 committed:
narrow_banner: Update the banners for deactivated users.

This commit updates the the banners for deactivated users in the
case of pm-with and group-pm narrows.

For the pm-with narrow, we display a banner for the deactivated user
that reads 'You have no direct messages with {person}.' and
shows the deactivated user's name in the banner with the subtitle
'{person} is no longer active in this organization.'.

For the group-pm narrow, we display a banner for the deactivated user
that reads 'You have no direct messages with these users.'
or 'You have no direct messages including {person}' depending on
the path through which the user visited the group-pm-with. The first
deactivated user's name is shown in the banner with the same subtitle.

Node tests were added accordingly.

Fixes #23408

palashb01 commented:
Update: I just fetched the latest changes and realized that a lot has changed in this section, with new tippy tooltips in the compose box and updated narrow banners. I am working on incorporating these latest updates and will push them by tomorrow

palashb01 commented:
Hey @alya  , a lot has changed in the past few days, including tippy tooltips and the DM-including operator, as well as the removal of group-PM-with. Therefore, I have updated the PR with the latest changes. Could you please review it, especially the narrow banners? Thanks :)

alya commented:
The screenshots look good, but I'm not seeing the tooltip on the disabled "Message..." button for some reason.

palashb01 commented:
@alya , could you please elaborate more? The tooltip appears with a delay, so if you hover over the button, the tooltip definitely appears.

![Screenshot from 2023-04-26 23-18-33](https://user-images.githubusercontent.com/66828942/234661828-917dec25-3148-4fb1-bc7c-ea8138e2474a.png)

alya commented:
I dunno, here's a very boring screen capture.

![2023-04-26 11 01 01](https://user-images.githubusercontent.com/2090066/234663824-9f409bae-04f7-4fc4-b217-e3a4dba80997.gif)

alya commented:
If I've recently deactivated the user, I'm also sometimes seeing the tooltip for the active button, even though the button is deactivated.

palashb01 commented:
That's strange. I couldn't reproduce the screenshot you attached. It is working fine for me. I am also attaching the live demo.

![pm-deactivated](https://user-images.githubusercontent.com/66828942/234664866-50a28a7e-ea82-4d94-ae79-e85310f53fb2.gif)

alya commented:
OK, I dunno, we have have @timabbott  take a look and see if he has any ideas for why we're seeing different things.
